### ISSUE TITLE
memory leak

### ISSUE THREAD

**Issue by zhouguangcai** [curl/curl#5545]
HI, Which version of OpenssL matches our current version? I used openssl 1.0.2u to compile the libcurl library, and compiled the configuration using DLL Debug - DLL openssl |Win32.There is a memory leak during the call.
### test code

	CURLcode resCode = curl_global_init(CURL_GLOBAL_ALL);
	CURL *curl = curl_easy_init();
	curl_easy_cleanup(curl);
	curl_global_cleanup(); 

### memory leak

```
Dumping objects ->
{4452} normal block at 0x0413CF78, 20 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
{4451} normal block at 0x0413CF28, 20 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
{4450} normal block at 0x0413CEE0, 12 bytes long.
 Data: < G          > F8 47 13 04 00 00 00 00 01 00 00 00 
{4449} normal block at 0x0413CE90, 16 bytes long.
 Data: <(   x           > 28 CF 13 04 78 CF 13 04 00 00 00 00 00 00 00 00 
{4448} normal block at 0x0413CE40, 20 bytes long.
 Data: <                > 02 00 00 00 90 CE 13 04 00 00 00 00 04 00 00 00 
{4447} normal block at 0x041347F8, 12 bytes long.
 Data: <    @       > 01 00 00 00 40 CE 13 04 02 00 00 00 
{3961} normal block at 0x04135010, 20 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
{3960} normal block at 0x04134FC8, 12 bytes long.
 Data: < N          > E0 4E 13 04 00 00 00 00 0E 00 00 00 
{3959} normal block at 0x04134F78, 16 bytes long.
 Data: < P              > 10 50 13 04 00 00 00 00 00 00 00 00 00 00 00 00 
{3958} normal block at 0x04134F28, 20 bytes long.
 Data: <    xO          > 01 00 00 00 78 4F 13 04 00 00 00 00 04 00 00 00 
{3957} normal block at 0x04134EE0, 12 bytes long.
 Data: <    (O      > 0E 00 00 00 28 4F 13 04 01 00 00 00 
{523} normal block at 0x04101070, 12 bytes long.
 Data: <            > 88 0F 10 04 00 00 00 00 00 00 00 00 
{522} normal block at 0x04101020, 16 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
{521} normal block at 0x04100FD0, 20 bytes long.
 Data: <                > 00 00 00 00 20 10 10 04 00 00 00 00 04 00 00 00 
{520} normal block at 0x04100F88, 12 bytes long.
 Data: <            > 00 00 00 00 D0 0F 10 04 00 00 00 00 
{497} normal block at 0x04100978, 12 bytes long.
 Data: <            > D8 B0 09 02 E0 CE 13 04 09 00 00 00 
{496} normal block at 0x04100928, 16 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
{495} normal block at 0x041008D8, 20 bytes long.
 Data: <    (           > 00 00 00 00 28 09 10 04 00 00 00 00 04 00 00 00 
{494} normal block at 0x0209B0D8, 12 bytes long.
 Data: <            > 09 00 00 00 D8 08 10 04 00 00 00 00 
{493} normal block at 0x04100858, 64 bytes long.
 Data: <p   x           > 70 10 10 04 78 09 10 04 00 00 00 00 00 00 00 00 
{492} normal block at 0x0209B038, 96 bytes long.
 Data: <X    vk  uk     > 58 08 10 04 10 76 6B 0F E0 75 6B 0F 08 00 00 00 
Object dump complete.
```

### curl/libcurl version
7-69-1 

### operating system
Windows- VS2010 

thanks

**Comment by bagder** [comment]
(that OpenSSL version is end-of-life'd, I suggest you upgrade)

I don't know how you determine that there's a leak, but I cannot reproduce any leaks on Linux using OpenSSL 1.0.2 with valgrind.

**Comment by jzakrzewski** [comment]
@bagder This is a Windows thing but not very smart. It uses _CrtDumpMemoryLeaks(). I have never managed to have this report clean except for hello world programs.

**Comment by jay** [comment]
If the leak is in curl source it should show the file and line number.

~~~
..\..\..\..\lib\curl_multibyte.c(67) : {3776} normal block at 0x003CD1F8, 3 bytes long.
 Data: <-v > 2D 76 00 
~~~

That is an actual leak fixed just now in f54b6c4. Without stack information it may be difficult to track down some of what it reports. It's of limited use. I get output similar to the reporter's. When there's no file/line then it is from some code that did not have the crt dbg enabled. This is most likely code outside of curl like OpenSSL.

The way to enable it is forced include crtdbg.h and then prepend to preprocessor `_CRTDBG_MAP_ALLOC;_CRT_SECURE_NO_WARNINGS;_CRT_NONSTDC_NO_DEPRECATE;`.

**Comment by zhouguangcai** [comment]
> (that OpenSSL version is end-of-life'd, I suggest you upgrade)
> 
> I don't know how you determine that there's a leak, but I cannot reproduce any leaks on Linux using OpenSSL 1.0.2 with valgrind.

I use on Windows, and if I upgrade OpensSL to 1.1.1 there will be more memory leaks

**Comment by zhouguangcai** [comment]
> If the leak is in curl source it should show the file and line number.
> 
> ```
> ..\..\..\..\lib\curl_multibyte.c(67) : {3776} normal block at 0x003CD1F8, 3 bytes long.
>  Data: <-v > 2D 76 00 
> ```
> 
> That is an actual leak fixed just now in [f54b6c4](https://github.com/curl/curl/commit/f54b6c4bc2f745fa32014819788c90126121729e). Without stack information it may be difficult to track down some of what it reports. It's of limited use. I get output similar to the reporter's. When there's no file/line then it is from some code that did not have the crt dbg enabled. This is most likely code outside of curl like OpenSSL.
> 
> The way to enable it is forced include crtdbg.h and then prepend to preprocessor `_CRTDBG_MAP_ALLOC;_CRT_SECURE_NO_WARNINGS;_CRT_NONSTDC_NO_DEPRECATE;`.

Yes, this is probably a leak in Openssl. If I just compile libcurl instead of using Openssl and make the same call, there won't be any memory leaks.So I would like to check if my Openssl version doesn't match

**Comment by jay** [comment]
> So I would like to check if my Openssl version doesn't match

I don't know what that means

**Comment by bagder** [comment]
> I upgrade OpensSL to 1.1.1 there will be more memory leaks

I don't believe that.

**Comment by zhouguangcai** [comment]
> > So I would like to check if my Openssl version doesn't match
> 
> I don't know what that means

want to confirm which version of Openssl I should be using

**Comment by zhouguangcai** [comment]
> > I upgrade OpensSL to 1.1.1 there will be more memory leaks
> 
> I don't believe that.

I don't believe it either, but it happened.If I use the 'DLL Debug 'compilation option that does not depend on OpensSL,  get the libcurl library and use it without memory leaks. If I use the 'DLL Debugg-DLL opensSL' compilation option,  get libcurl and use it in VS, VS prompts for memory leaks.
Did I compile the configuration problem?
My test code just calls initialization and cleanup

```
CURLcode resCode = curl_global_init(CURL_GLOBAL_ALL);
CURL *curl = curl_easy_init();
curl_easy_cleanup(curl);
curl_global_cleanup(); 
```

**Comment by bagder** [comment]
I don't know what you do. We use OpenSSL 1.1.1 in many setups all the time and we don't see any leaks when entire test suite is run on multiple platforms.

If you think libcurl leaks memory, please tell us what allocation libcurl did that wasn't freed at exit. I cannot reproduce this on my machines, but I also don't run this on Windows.

**Comment by zhouguangcai** [comment]
I debugged the source code. The licurl library called the OPENSSL_load_builtin_modules and ENGINE_load_builtin_engines interfaces during initialization. There was no corresponding cleanup when exiting. I actively call CONF_modules_unload(1) and
   OPENSSL_cleanup() interface, there will be no memory leak at this time. The official Openssl documentation does state that we do not need to call the cleanup interface, and automatically clean up related resources when exiting. I used visual studio 2010. I don’t know why.
Thinks and Good luck
